### PR TITLE
[TIMOB-18062] iOS: AttributedString Parity with Android

### PR DESCRIPTION
**Pull request by ashcoding**
Jira: https://jira.appcelerator.org/browse/TIMOB-18062

This PR is a continuation of https://github.com/appcelerator/titanium_mobile/pull/6409/ which came from #6358.

It addresses the comments by @pingwang2011 for the Android portion of the code. The doc is still not completed yet as Titanium.UI.iOS.Attribute and Titanium.UI.iOS.AttributedString yml files are missing the "deprecated".

The propertyAccessors is now used in AttributedStringProxy.java, AttributeProxy.java and LabelProxy.java.

### CODE REVIEW

File: android/modules/ui/src/java/ti/modules/titanium/ui/widget/TiUILabel.java [hunk 1]
```diff
@@ -276,6 +286,11 @@ public void propertyChanged(String key, Object oldValue, Object newValue, KrollP
 		} else if (key.equals(TiC.PROPERTY_SHADOW_COLOR)) {
 			shadowColor = TiConvert.toColor(TiConvert.toString(newValue));
 			tv.setShadowLayer(shadowRadius, shadowX, shadowY, shadowColor);
+		} else if (key.equals(TiC.PROPERTY_ATTRIBUTED_STRING) && newValue instanceof AttributedStringProxy) {
```
Comment (pingwang2011) [inline android/modules/ui/src/java/ti/modules/titanium/ui/widget/TiUILabel.java]: @bhatfield, @vishalduggal,  should we resolve this parity issue or just document it?

**Comment by pingwang2011** [comment]
Code reviewed. Left one comment.
Ran the above sample code. It works as expected.

File: android/modules/ui/src/java/ti/modules/titanium/ui/widget/TiUILabel.java (hunk 1 above)
Comment (negupta) [inline android/modules/ui/src/java/ti/modules/titanium/ui/widget/TiUILabel.java]: We should try to resolve the parity issue if we can.

| Neeraj Gupta | SVP, Product and Engineering
| Appcelerator, Inc | e: ngupta@appcelerator.com<mailto:%20ngupta@appcelerator.com> | o:+1 650-528-2931 |
| 440 N. Bernardo Ave, Mtn View, CA 94043

**Comment by ingo** [comment]
@ashcoding can you finish the review?

**Comment by ashcoding** [comment]
Will do.

**Comment by ashcoding** [comment]
@pingwang2011 With regards to the comment on parity you left, is the current PR okay? Or is there something that is expected to be added/changed with regards to parity? 

Besides that, unless I miss out on something, there's nothing else left unaddressed.

**Comment by ashcoding** [comment]
iOS code needs to change for the color property. Sample code to show issue as follows:
```
// this sets the background color of the master UIView (when there are no windows/tab groups on it)
Titanium.UI.setBackgroundColor('#000');
//////////////////////////////////////
var win1 = Titanium.UI.createWindow({  
    title:'Label',
    backgroundColor:'#ddd'
});

var text =  'Bacon ipsum dolor Appcelerator Titanium rocks! sit amet fatback leberkas salami sausage tongue strip steak.';

var attr = Ti.UI.createAttributedString({
    text: text,
    attributes: [
        //font
        {
            type: Ti.UI.ATTRIBUTE_FONT,
            value: {font:{fontSize:50,fontFamily:'Helvetica Neue'}},
            range: [0, text.length]
        },
        // Underlines text
        {
            type: Ti.UI.ATTRIBUTE_UNDERLINES_STYLE,
            value: Ti.UI.ATTRIBUTE_UNDERLINE_STYLE_SINGLE, //only iOS needs this, android ignores.
            range: [0, text.length]
        },
        // Sets a background color
        {
            type: Ti.UI.ATTRIBUTE_BACKGROUND_COLOR,
            value: "red",
            range: [text.indexOf('Appcelerator'), ('Appcelerator').length]
        },
        
        {
            type: Ti.UI.ATTRIBUTE_BACKGROUND_COLOR,
            value: "blue",
            range: [text.indexOf('Titanium'), ('Titanium').length]
        },
        
        {
            type: Ti.UI.ATTRIBUTE_BACKGROUND_COLOR,
            value: "yellow",
            range: [text.indexOf('rocks!'), ('rocks!').length]
        },
        
        {
            type: Ti.UI.ATTRIBUTE_FOREGROUND_COLOR,
            value: "orange",
            range: [0,  text.length]
        },
        
        {
            type: Ti.UI.ATTRIBUTE_FOREGROUND_COLOR,
            value: "black",
            range: [text.indexOf('rocks!'), ('rocks!').length]
        },
        
        {
            type: Ti.UI.ATTRIBUTE_LINK,//ignored in iOS, label doesn't support attribute link. Use textArea instead. 
            value: "http://www.appcelerator.com",
            range: [text.indexOf('Appcelerator'), ('Appcelerator').length]
        }

    ]
});


var label = Titanium.UI.createLabel({
    left: 20,
    right: 20,
    height: Titanium.UI.SIZE,
    attributedString: attr
});

//Changing the following properties shouldn't affect label as attributedString is respected first
label.text = "New text is here"; // This doesn't change on iOS and Android. Expected behavior.
label.setFont({fontSize:5}); // This doesn't change on iOS and Android. Expected behavior.
label.color = "red"; //iOS seems to be affected by this. Unexpected behavior.
//Shadow properties are working in iOS and Android. Doc should mention this.
label.shadowColor = "blue"; 
label.shadowRadius = 5;

win1.add(label);
win1.open();
```

**Comment by cheekiatng** [comment]
After some verification, existing iOS "attributedString"' behavior is DIFFERENT from what the existing documents say.
if "attributedString" and other properties of UILabel (or other UI components involving text) is set initially, iOS respects the other properties. This has been confirmed. So there should no further changes to IOS behavior, and we will require Android to achieve parity with this behavior. Documents must also be updated to inform the correct behavior.
Thanks Ash.

**Comment by ashcoding** [comment]
I'll go change the Android behavior of the attributedString to match iOS when other properties is set.

**Comment by ashcoding** [comment]
This should be checked with this now:
```
// this sets the background color of the master UIView (when there are no windows/tab groups on it)
Titanium.UI.setBackgroundColor('#000');
//////////////////////////////////////
var win1 = Titanium.UI.createWindow({  
    title:'Label',
    backgroundColor:'#ddd'
});

var text =  'Bacon ipsum dolor Appcelerator Titanium rocks! sit amet fatback leberkas salami sausage tongue strip steak.';

var attr = Ti.UI.createAttributedString({
    text: text,
    attributes: [
        //font
        {
            type: Ti.UI.ATTRIBUTE_FONT,
            value: {font:{fontSize:50,fontFamily:'Helvetica Neue'}},
            range: [0, text.length]
        },
        // Underlines text
        {
            type: Ti.UI.ATTRIBUTE_UNDERLINES_STYLE,
            value: Ti.UI.ATTRIBUTE_UNDERLINE_STYLE_SINGLE, //only iOS needs this, android ignores.
            range: [0, text.length]
        },
        // Sets a background color
        {
            type: Ti.UI.ATTRIBUTE_BACKGROUND_COLOR,
            value: "red",
            range: [text.indexOf('Appcelerator'), ('Appcelerator').length]
        },

        {
            type: Ti.UI.ATTRIBUTE_BACKGROUND_COLOR,
            value: "blue",
            range: [text.indexOf('Titanium'), ('Titanium').length]
        },

        {
            type: Ti.UI.ATTRIBUTE_BACKGROUND_COLOR,
            value: "yellow",
            range: [text.indexOf('rocks!'), ('rocks!').length]
        },

        {
            type: Ti.UI.ATTRIBUTE_FOREGROUND_COLOR,
            value: "orange",
            range: [0,  text.length]
        },

        {
            type: Ti.UI.ATTRIBUTE_FOREGROUND_COLOR,
            value: "black",
            range: [text.indexOf('rocks!'), ('rocks!').length]
        },

        {
            type: Ti.UI.ATTRIBUTE_LINK,//ignored in iOS, label doesn't support attribute link. Use textArea instead. 
            value: "http://www.appcelerator.com",
            range: [text.indexOf('Appcelerator'), ('Appcelerator').length]
        }

    ]
});


var label = Titanium.UI.createLabel({
    left: 20,
    right: 20,
    height: Titanium.UI.SIZE,
    attributedString: attr
});

//This will affect and change the label. Android and iOS functions the same
label.text = "New text is here";
label.setFont({fontSize:5}); 
label.color = "red"; 
label.shadowColor = "blue"; 
label.shadowRadius = 5;

win1.add(label);
win1.open();
```

**Comment by ashcoding** [comment]
I looked into this further.

In Android, if you set an attributed string, when you use methods such as setFont or setColor, natively it sets it to the Text in the TextView to those properties. If the AttributedString (natively it is SpannableString), those methods do not seem to change the properties of the AttributedString. 

setColor, setFont and etc would affect the whole Text when used normally while normally when you use AttributedStrings, you would specify at which part does it have a certain color. Thus, I don't think it should be used in that way too.

So in this case, I would not recommend for parity in this case when AttributedString, text, color and font is used. 

Instead I would recommend a document change to specify that you should not use AttributedString with the other properties, eg text, font and color in Android. Either use AttributedString or use the other properties.

Shadow properties work with no issue.

**Comment by pingwang2011** [comment]
CR & FR passed for Android.
Please attach the test case to the JIRA ticket so QE can test it.

**Comment by vishalduggal** [comment]
I am going to accept this PR as is. There are some problems with the iOS implementation (formatting, un-required checks etc) but that will be fixed in a subsequent PR.

__APPROVED__

**Comment by alexgbernier** [comment]
Custom fonts don't appear to be working: https://jira.appcelerator.org/browse/AC-70

**Pull request merged**